cpp/src/iterator/iterator.cuh

```diff
@@ -0,0 +1,406 @@
+template<typename T_element, bool has_nulls>
+struct column_input_pair;
+
+// @overload column_input_pair<T_element, false>
+template<typename T_element>
+struct column_input_pair<T_element, false> : public column_input<T_element, false>
+{
+    CUDA_HOST_DEVICE_CALLABLE
+    column_input_pair(const T_element *_data, const bit_mask::bit_mask_t *_valid, T_element _identity)
+    : column_input<T_element, false>(_data, _valid, _identity){};
+
+    CUDA_HOST_DEVICE_CALLABLE
+    auto at(gdf_index_type id) const {
+        return thrust::make_pair(this->data[id], true);
+    };
+
+};
+
+// @overload column_input_pair<T_element, true>
+template<typename T_element>
+struct column_input_pair<T_element, true> : public column_input<T_element, true>
+{
+    CUDA_HOST_DEVICE_CALLABLE
+    column_input_pair(const T_element *_data, const bit_mask::bit_mask_t *_valid, T_element _identity)
+    : column_input<T_element, true>(_data, _valid, _identity){};
+
+    CUDA_HOST_DEVICE_CALLABLE
+    auto at(gdf_index_type id) const {
+        return this->is_valid(id) ?
+            thrust::make_pair(this->data[id], true):
+            thrust::make_pair(this->identity, false);
+    };
+};
+
+/** -------------------------------------------------------------------------*
+ * @brief column input iterator to support null bitmask
+ * The input iterator which can be used for cub and thrust.
+ * This is derived from `thrust::iterator_adaptor`
+ * T_column_input = column_input or column_input_pair
+ * T_column_input will provide the value at index `id` with/without null bitmask.
+ *
+ * @tparam  T_iterator_output The output data value type of the iterator
+ * @tparam  T_column_input  The input struct type of column_input or column_input_pair
+ * @tparam  Iterator The base iterator which gives the index of array.
+ *                   The default is `thrust::counting_iterator`
+ * -------------------------------------------------------------------------**/
+template<typename T_iterator_output, typename T_column_input,
+    typename Iterator=thrust::counting_iterator<gdf_index_type> >
+  class column_input_iterator
+    : public thrust::iterator_adaptor<
+        column_input_iterator<T_iterator_output, T_column_input, Iterator>, // the name of the iterator we're creating
+        Iterator,                   // the name of the iterator we're adapting
+        T_iterator_output,          // set `super_t::value` to `T_iterator_output`
+        thrust::use_default, thrust::use_default,
+        T_iterator_output,          // set `super_t::reference` to `T_iterator_output`
+        thrust::use_default
+      >
+  {
+  public:
+    // shorthand for the name of the iterator_adaptor we're deriving from
+    using super_t = thrust::iterator_adaptor<
+      column_input_iterator<T_iterator_output, T_column_input, Iterator>, Iterator,
+      T_iterator_output, thrust::use_default, thrust::use_default, T_iterator_output, thrust::use_default
```

Comment:
Like Michal said, I think you should leave the 6th template argument as the default. Since it will default to `T_iterator_output&`.